### DISCUSSION TITLE
Can't pass checkMesh after several optimisation cycles

### DISCUSSION THREAD

**Discussion by Otherside5** [mdolab/dafoam#144] [General]
Hello everyone,

I am simulating a centrifugal compressor with DATurboFoam with a 140k cell unstructured mesh.

For the first 6 optimisation iterations everything runs smoothly and then it seems like he is not able to find a mesh without severe errors and after some time it stops. The error log is attached. I am wondering if there is anything i can do to prevent this from happening or if my simulation/mesh is just not setup correctly.

[error_log.txt](https://github.com/mdolab/dafoam/files/6604169/error_log.txt)

Also it seems like he can not find the outlet patch faces, which is odd since he can find the faces of all other patches.
"Found matching patch outlet with 0 faces."
Should I be worried about this?

Thanks in advance
Alex

**Comment by friedenhe** [comment]
Which optimizer you are using? We found that the SNOPT optimizer worked best for compressor cases because it has a mechanism to automatically reduce the optimization step size when mesh quality fails. SLSQP and IPOPT do not have this mechanism and will keep trying poor mesh quality designs. But even with SNOPT, compressor cases can still have mesh quality issues because it is an internal flow case. Another way you may want to check is to lower the upper and lower bounds of the design variables (e.g., from -2e-3 to 2e-3). We recently updated the DAFoam version and the optimization will output the bad quality mesh to the postProcessing folder. You can load them in Paraview and figure out which part of the mesh is failing, and then reduce the design variable bounds or adjust the mesh in that area. Let me know if this helps.

**Reply by Otherside5** [reply]
Thank you for your answer!

I am currently using SLSQP, I have also tried it with IPOPT.

When I try to use SNOPT it crashes:
"pyOptSparse Error: There was an error importing the compiled snopt module"

Do I have to do something before using SNOPT? 
(I use the Docker image, this happens with v2.5.5 and also the version I used before)

I am currently optimising with lower bounds on the design variables and will post an update once the simulation is done.

**Reply by friedenhe** [reply]
SNOPT is commercial software and the DAFoam docker image comes without SNOPT. Check [here](https://mdolab-pyoptsparse.readthedocs-hosted.com/en/latest/optimizers/SNOPT.html) for more details.

**Reply by Otherside5** [reply]
Good to know, but unfortunately that is not an option at the moment.

I tried changing the design variable bounds but it did not change anything (I tried bigger and smaller bounds)

I suspect that the problem is that two walls are nearly touching each other. Everytime I simulate the blade comes closer and closer to the outer wall. 
![comp](https://user-images.githubusercontent.com/83928129/121581760-872b4b80-ca2e-11eb-98e3-53f49d76fde0.PNG)

My second guess would be that the mesh at the boundary from blade to inner wall is to bad, that is also the boundary between design surface and the rest of the mesh.

How can I save the failed meshes in DAFoam or should that happen automatically?

**Reply by friedenhe** [reply]
Yes, it will be tricky to deal with mesh quality for an internal flow case. As a first try, I suggest you fix the FFD movement near the tip and root. If you want the tip to move, you may need to set a thickness constraint for the tip clearance. Otherwise, the optimizer will always try to close the gap. To set such constraint, you need to define curved leList and teList with multiple points. Regarding the mesh quality at the root, one way to fix it is to just fix the FFD at the root. Another way is to set the hub as the design surface such that when the FFD points move, the hub and blade will move together, which should alleviate the mesh quality issue to some extent.
